tests/test_dot.py

```diff
@@ -0,0 +1,114 @@
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import os
+import tempfile
+import unittest
+
+import retworkx
+
+
+class TestDot(unittest.TestCase):
+    def setUp(self):
+        fd, self.path = tempfile.mkstemp()
+        os.close(fd)
+        os.remove(self.path)
+
+    def test_graph_to_dot(self):
+        graph = retworkx.PyGraph()
+        graph.add_node({'color': 'black', 'fillcolor': 'green',
+                        'label': "a", 'style': 'filled'})
+        graph.add_node({'color': 'black', 'fillcolor': 'red',
+                        'label': "a", 'style': 'filled'})
+        graph.add_edge(0, 1, dict(label='1', name='1'))
+        expected = (
+            "graph {\n0 [color=black, fillcolor=green, label=a, style=filled"
+            "];\n1 [color=black, fillcolor=red, label=a, style=filled];"
+            "\n0 -- 1 [label=1, name=1];\n}\n")
+        res = graph.to_dot(lambda node: node, lambda edge: edge).decode('utf8')
+        self.assertEqual(expected, res)
+
+    def test_digraph_to_dot(self):
+        graph = retworkx.PyDiGraph()
+        graph.add_node({'color': 'black', 'fillcolor': 'green',
+                        'label': "a", 'style': 'filled'})
+        graph.add_node({'color': 'black', 'fillcolor': 'red',
+                        'label': "a", 'style': 'filled'})
+        graph.add_edge(0, 1, dict(label='1', name='1'))
+        expected = (
+            "digraph {\n0 [color=black, fillcolor=green, label=a, style=filled"
+            "];\n1 [color=black, fillcolor=red, label=a, style=filled];"
+            "\n0 -> 1 [label=1, name=1];\n}\n")
+        res = graph.to_dot(lambda node: node, lambda edge: edge)
```

Comment:
this could be replaced so that node data is not the same as dot attribute options, right?

Reply:
Right, that was my reasoning behind making them callables. The node data could be anything so you could define functions to adjust the formatting based on the data. A simple example would be something like:
```python3
import retworkx
def node_color(data):
    if data > 4:
        return {'color': 'black'}
    else:
        return {'color': 'red'}
graph = retworkx.PyDiGraph
a = graph.add_node(52)
graph.add_child(a, 2, {})
graph.to_dot(node_color, lambda edge: edge)
```

What I was thinking of was for being able to replace https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/visualization/dag_visualization.py#L97-L111 with a function that could be passed to the dot method here.